### ISSUE TITLE
SmolLM2-360M Fine-Tuning on DocVQA Results in NaN Loss

### ISSUE THREAD

**Issue by dsnsabari** [huggingface/nanoVLM#50]
I am trying to fine-tune HuggingFaceTB/SmolLM2-360M on the DocVQA dataset, but I'm encountering NaN values in the training loss. I am training the model in Colab. Please find my configuration below.

```
@dataclass
class VLMConfig:
    vit_hidden_dim: int = 768
    vit_inter_dim: int = 4 * vit_hidden_dim
    vit_patch_size: int = 16
    vit_img_size: int = 224
    vit_n_heads: int = 12
    vit_dropout: float = 0.0
    vit_n_blocks: int = 12
    vit_ln_eps: float = 1e-6
    vit_cls_flag: bool = False
    vit_model_type: str = 'google/siglip-base-patch16-224'

    lm_hidden_dim: int = 1024 #576
    lm_inter_dim: int = 4096 #1536
    lm_rms_eps: float = 1e-5
    lm_re_base: int = 100000
    lm_max_position_embeddings: int = 8192
    lm_vocab_size: int = 49152
    lm_n_heads: int = 16 #9
    lm_n_kv_heads: int = 4 #3
    lm_dropout: float = 0.0
    lm_n_blocks: int = 24 # 30
    lm_attn_scaling: float = 1.0
    lm_eos_token_id: int = 0
    lm_max_length: int = 128 #128 - 49  # Deduct the image token lenght to achieve a 'nice number'
    lm_use_tokens: bool = False # Decide if the LM expects tokens or embeddings as input (if using as a backbone for the VLM, set to False)
    lm_tie_weights: bool = True # Decide if you want to tie the LM Head weight to the token embeding weights
    lm_model_type: str = 'HuggingFaceTB/SmolLM2-360M'
    lm_tokenizer: str = 'HuggingFaceTB/cosmo2-tokenizer'

    mp_pixel_shuffle_factor: int = 2

    vlm_load_backbone_weights: bool = True
    vlm_checkpoint_path: str = 'checkpoints/nanoVLM-222M'


@dataclass
class TrainConfig:
    lr_mp: float = 1e-3
    lr_backbones: float = 5e-5
    compile: bool = False
    data_cutoff_idx: int = 1024 # Let's only use a small subset of the data at first, otherwise it takes very long to see anything :D
    batch_size: int = 12
    mmstar_batch_size: int = 12
    epochs: int = 15
    eval_in_epochs: bool = False # Deactivating this in colab, because it would evaluate 1500 samples of MMStar every time otherwise
    resume_from_vlm_checkpoint: bool = False # Indicate if the training should be resumed from a checkpoint of the whole VLM or you want to start from scratch
    train_dataset_path: str = 'HuggingFaceM4/the_cauldron'
    train_dataset_name: tuple[str, ...] = ("docvqa",) #("tqa", "vsr") #All options; ("ai2d", "aokvqa", "chart2text", "chartqa", "clevr", "cocoqa", "datikz", "diagram_image_to_text", "docvqa", "dvqa", "figureqa", "finqa", "geomverse", "hateful_memes", "hitab", "iam", "iconqa", "infographic_vqa", "intergps", "localized_narratives", "mapqa", "multihiertt", "ocrvqa", "plotqa", "raven", "rendered_text", "robut_sqa", "robut_wikisql", "robut_wtq", "scienceqa", "screen2words", "st_vqa", "tabmwp", "tallyqa", "tat_qa", "textcaps", "textvqa", "tqa", "vistext", "visual7w", "visualmrc", "vqarad", "vqav2", "vsr", "websight") # "clevr_math", "okvqa", "spot_the_diff", "nlvr2", "mimic_cgd",
    test_dataset_path: str = "Lin-Chen/MMStar"
```

```
100%|██████████| 85/85 [01:20<00:00,  1.05it/s]
Epoch 1/15, Train Loss: nan | Time: 80.98s | T/s: 994.62
100%|██████████| 85/85 [01:20<00:00,  1.06it/s]
Epoch 2/15, Train Loss: nan | Time: 80.20s | T/s: 1003.94
100%|██████████| 85/85 [01:21<00:00,  1.04it/s]
Epoch 3/15, Train Loss: nan | Time: 81.87s | T/s: 983.93
100%|██████████| 85/85 [01:19<00:00,  1.07it/s]
Epoch 4/15, Train Loss: nan | Time: 79.50s | T/s: 1012.89
100%|██████████| 85/85 [01:18<00:00,  1.09it/s]
Epoch 5/15, Train Loss: nan | Time: 78.28s | T/s: 1028.72
100%|██████████| 85/85 [01:17<00:00,  1.09it/s]
Epoch 6/15, Train Loss: nan | Time: 77.89s | T/s: 1034.11
 41%|████      | 35/85 [00:32<00:44,  1.12it/s]

```

**Comment by andimarafioti** [comment]
That's weird, it shouldn't give you nans. Maybe the data is not being parsed properly for docvqa?

**Comment by dsnsabari** [comment]
@andimarafioti, but when I use the same dataset to train the 'HuggingFaceTB/SmolLM2-135M', it does not give the Nans.

**Comment by lusxvr** [comment]
Interesting, have you tried increasing the `lm_max_lenght`? DocVQA has quite long sequences and the collator currently just skips samples that are longer than the max length. So if every sample is skipped this could result in a nan for the loss.

**Comment by dsnsabari** [comment]
@lusxvr @andimarafioti  Even after changing the dataset to ('tqa', 'vsr') and increasing lm_max_length to 256, I am still getting NaNs in the training loss.    

``` 
100%|██████████| 85/85 [00:46<00:00,  1.83it/s]
Epoch 1/15, Train Loss: nan | Time: 46.48s | T/s: 2260.22
100%|██████████| 85/85 [00:45<00:00,  1.88it/s]
Epoch 2/15, Train Loss: nan | Time: 45.33s | T/s: 2317.66
 24%|██▎       | 20/85 [00:11<00:37,  1.75it/s] 

```

**Comment by dsnsabari** [comment]
@andimarafioti @lusxvr  When I train the 'HuggingFaceTB/SmolLM2-135M' model with a modified lm_max_length and the parameters below, I am getting NaN in the training loss. Could you help me resolve this issue?

```

@dataclass
class VLMConfig:
    vit_hidden_dim: int = 768
    vit_inter_dim: int = 4 * vit_hidden_dim
    vit_patch_size: int = 16
    vit_img_size: int = 224
    vit_n_heads: int = 12
    vit_dropout: float = 0.0
    vit_n_blocks: int = 12
    vit_ln_eps: float = 1e-6
    vit_cls_flag: bool = False
    vit_model_type: str = 'google/siglip-base-patch16-224'

    lm_hidden_dim: int = 576 #576
    lm_inter_dim: int = 2048 #1536
    lm_rms_eps: float = 1e-5
    lm_re_base: int = 100000
    lm_max_position_embeddings: int = 8192
    lm_vocab_size: int = 49152
    lm_n_heads: int = 16 #9
    lm_n_kv_heads: int = 4 #3
    lm_dropout: float = 0.0
    lm_n_blocks: int = 24 # 30
    lm_attn_scaling: float = 1.0
    lm_eos_token_id: int = 0
    lm_max_length: int = 256 #128 - 49  # Deduct the image token lenght to achieve a 'nice number'
    lm_use_tokens: bool = False # Decide if the LM expects tokens or embeddings as input (if using as a backbone for the VLM, set to False)
    lm_tie_weights: bool = True # Decide if you want to tie the LM Head weight to the token embeding weights
    lm_model_type: str = 'HuggingFaceTB/SmolLM2-135M' #'HuggingFaceTB/SmolLM2-360M'
    lm_tokenizer: str = 'HuggingFaceTB/cosmo2-tokenizer'

    mp_pixel_shuffle_factor: int = 2

    vlm_load_backbone_weights: bool = True
    vlm_checkpoint_path: str = 'checkpoints/nanoVLM-222M'


@dataclass
class TrainConfig:
    lr_mp: float = 1e-3
    lr_backbones: float = 5e-5
    compile: bool = False
    data_cutoff_idx: int = 1024 # Let's only use a small subset of the data at first, otherwise it takes very long to see anything :D
    batch_size: int = 12
    mmstar_batch_size: int = 12
    epochs: int = 15
    eval_in_epochs: bool = False # Deactivating this in colab, because it would evaluate 1500 samples of MMStar every time otherwise
    resume_from_vlm_checkpoint: bool = False # Indicate if the training should be resumed from a checkpoint of the whole VLM or you want to start from scratch
    train_dataset_path: str = 'HuggingFaceM4/the_cauldron'
    train_dataset_name: tuple[str, ...] = ("docvqa", "vsr") #("docvqa",) #("tqa", "vsr") #All options; ("ai2d", "aokvqa", "chart2text", "chartqa", "clevr", "cocoqa", "datikz", "diagram_image_to_text", "docvqa", "dvqa", "figureqa", "finqa", "geomverse", "hateful_memes", "hitab", "iam", "iconqa", "infographic_vqa", "intergps", "localized_narratives", "mapqa", "multihiertt", "ocrvqa", "plotqa", "raven", "rendered_text", "robut_sqa", "robut_wikisql", "robut_wtq", "scienceqa", "screen2words", "st_vqa", "tabmwp", "tallyqa", "tat_qa", "textcaps", "textvqa", "tqa", "vistext", "visual7w", "visualmrc", "vqarad", "vqav2", "vsr", "websight") # "clevr_math", "okvqa", "spot_the_diff", "nlvr2", "mimic_cgd",
    test_dataset_path: str = "Lin-Chen/MMStar"


```

**Comment by geronimi73** [comment]
@dsnsabari precision issue. change the dtype to `bfloat16`.

This is the line you have to change.
From
```python
            with torch.autocast(device_type='cuda', dtype=torch.float16): # Mixed precision training
```
To
```python
            with torch.autocast(device_type='cuda', dtype=torch.bfloat16): # Mixed precision training
```

It won't run on a free Google Colab notebook though, bfloat16 is not supported by T4 GPUs. You can rent a prev. generation GPU like a 3090 for a few cents per hour in the cloud.

**Alternatively**, if you want to stay on Google Colab you could train in `float32` by removing the line `with torch.autocast`. This will need more VRAM and force you to reduce your batch size, try `batch_size: int = 6`.

**Comment by dsnsabari** [comment]
@geronimi73 Thank you so much. It is working .
